### PR TITLE
SubNav v2

### PR DESCRIPTION
**Pull request by rezrah**
## Summary

Part of https://github.com/github/primer/issues/3312

Updates `SubNav` component to support anchor-based navigation similar to `AnchorNav`.

⚠️ Despite being internally referred to as `SubNav v2` there are no breaking changes to the API as part of this update. A full visual refresh however, has been implemented per latest Site guidance.

🔗  [Documentation](https://primer-dbe99eb951-26139705.drafts.github.io/brand/components/SubNav#optional-anchor-based-sub-menu)
🔗  [Storybook examples](https://primer-dbe99eb951-26139705.drafts.github.io/brand/storybook/?path=/story/components-subnav-features--dropdown-variant)

## List of notable changes:

- updated `SubNav.Submenu` to support `variant="anchor"` 
- updated Storybook examples to respect color modes and differentiate between dropdown and anchor submenu types
- fixed various bugs in existing subnav like not closing mobile overlays when clicking outside it

## What should reviewers focus on?

- Manually review each of the Storybook examples for design accuracy
- Read through the docs and verify new features are accurately captured

## Steps to test:

1. Use [these Storybook examples](https://primer-dbe99eb951-26139705.drafts.github.io/brand/storybook/?path=/story/components-subnav-features--dropdown-variant)
1. Use [this docs link](https://primer-dbe99eb951-26139705.drafts.github.io/brand/components/SubNav#optional-anchor-based-sub-menu)
1. Use the visual diffs to see before and afters

## Supporting resources (related issues, external links, etc):

- [Tracking issue](https://github.com/github/primer/issues/3312)

## Contributor checklist:

- [x] All new and existing CI checks pass
- [x] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [x] UI Changes contain new visual snapshots (generated by adding `update snapshots` label to the PR)
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist:

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

## Screenshots:

> Please try to provide before and after screenshots or videos

<table>
<tr>
<th> Before</th> <th> After </th>
</tr>
<tr>
<td valign="top">

image

 </td>
<td valign="top">

image

</td>
</tr>
</table>

New anchor navigation

<table>
<tr>
<th> Light</th> <th> Dark </th>
</tr>
<tr>
<td valign="top">

Screenshot 2024-11-26 at 10 49 38

 </td>
<td valign="top">

Screenshot 2024-11-26 at 10 49 44

</td>
</tr>
</table>

### CODE REVIEW

File: packages/react/src/SubNav/SubNav.tsx
Comment (joshfarrant) [inline packages/react/src/SubNav/SubNav.tsx:399]: Unsure why, but this doesn't seem to be working for me in Chrome on Windows. Hitting space or Enter while the button is focussed doesn't do anything.

Happy to do a bit more digging, but I wanted to flag now so I didn't forget.

File: packages/react/src/SubNav/SubNav.visual.spec.ts-snapshots/Visual-Comparison-SubNav-Mobile-viewport-test-e0f45-en-SubNav-Narrow-Dropdown-Variant-Menu-Open-1-linux.png
Comment (joshfarrant) [inline packages/react/src/SubNav/SubNav.visual.spec.ts-snapshots/Visual-Comparison-SubNav-Mobile-viewport-test-e0f45-en-SubNav-Narrow-Dropdown-Variant-Menu-Open-1-linux.png:1]: The spacing around "Copilot" feels a little off here. Unsure whether it's because it's the active item, or if it's due to it being the parent of a list.

File: packages/react/src/SubNav/SubNav.tsx
Comment (joshfarrant) [inline packages/react/src/SubNav/SubNav.tsx:471]: I'm not sure what the ideal behaviour is here, but at the moment activating an anchor nav link scrolls the page but doesn't move focus. This mean that a screen reader user, or a keyboard user, can activate a link however when they try to navigate to the next element they will still be inside this list of anchor links, which is counter-intuitive.

I'd imagine that we'd also want to move focus to the element that the anchor nav is linking to, wdyt?

**Review by joshfarrant — CHANGES_REQUESTED** [review]

**Review by joshfarrant — COMMENTED** [review]
Commenting to remove _Changes requested_ status

File: packages/react/src/SubNav/SubNav.tsx
Comment (joshfarrant) [reply packages/react/src/SubNav/SubNav.tsx:399]: Ah ok, this is a new one for me. The `onKeyDown` is working on Windows, however the screen reader I'm using (NVDA) is eating the keypresses when in browse mode. If I switch to focus mode (the mode used for interacting with the page) then it works as expected.

It could be easy to dismiss this as unintuitive screen-reader behaviour, but from a user point of view they just hear that they're focussing a button (_"Open submenu button collapsed"_) but they're unable to open it using space or Enter, which should work for a button.

I'm guessing they open/close behaviour is bound to keydown here to prevent users from being able to click on the dropdown button? My initial feeling is that the simplest way forward would be to remove that logic and treat this like a normal button and bind the open/close behaviour to a click instead.

wdyt?

File: packages/react/CHANGELOG.md
Comment (rezrah) [reply packages/react/CHANGELOG.md:595]: 👍 it's intentional

File: packages/react/src/SubNav/SubNav.tsx
Comment (rezrah) [reply packages/react/src/SubNav/SubNav.tsx:277]: I recall us receiving guidance from a11y design team on this being okay, that it should have an aria-label to provide better context on what users should do here rather than showing an arbitrary value like "Overview". Has that guidance changed?

Comment (joshfarrant) [reply packages/react/src/SubNav/SubNav.tsx:277]: If we've had that guidance before then I'd say stick to that. It feels unintuitive to me because I can see something on the screen that I can't have the screen reader read out to me.

I'll reach out for guidance on this so I know for future reference, but don't let this block this PR

File: packages/react/src/SubNav/SubNav.tsx
Comment (rezrah) [reply packages/react/src/SubNav/SubNav.tsx:471]: Thanks 👍, I've removed the `scrollTo` and reverted to native browser behaviour to resolve this. Now works the same as AnchorNav, which should be correct.

File: packages/react/src/SubNav/SubNav.tsx
Comment (rezrah) [reply packages/react/src/SubNav/SubNav.tsx]: For posterity, chatted through this with @joshfarrant. Memoization of children specifically removed, and retained for other props.

File: packages/react/src/SubNav/SubNav.visual.spec.ts-snapshots/Visual-Comparison-SubNav-Mobile-viewport-test-e0f45-en-SubNav-Narrow-Dropdown-Variant-Menu-Open-1-linux.png
Comment (rezrah) [reply packages/react/src/SubNav/SubNav.visual.spec.ts-snapshots/Visual-Comparison-SubNav-Mobile-viewport-test-e0f45-en-SubNav-Narrow-Dropdown-Variant-Menu-Open-1-linux.png:1]: Thanks, there was some extra padding being applied to sub-menu.. added a new rule to remove that.

**Review by joshfarrant — APPROVED** [review]
Looking great! :shipit:

**Comment by danielguillan** [comment]
This looks really great!

Just a few minor visual adjustments needed:

- Heading font size:
  - small: `20px` → `16px`  
  - large: `20px` → `18px`
- SubNav Links font size: 
  - large: `16px` → `14px`
- Gap between links: 
  - large: `24px` → `20px`
- Separator (`/`) color: 
  - light mode: `gray-4`
  - dark mode: `gray-5`

AnchorNav top and bottom paddings should be set to `12` and `20` respectively. We might need to adjust the link line-height and/or paddings accordingly to compensate the overall height.

image

Dropdown items should align with the parent link above.
screenshot-OLT5Vo8h-000053@2x

[Link to the Figma implementation](https://www.figma.com/design/BJ95AjraesmRCWsKA013GS/branch/phPNHrO44XjnnFqnA17Aza/Primer-Brand?node-id=12997-5682&node-type=canvas&m=dev) for reference.

I'll preemptively approve so you can merge when these tweaks are ready.

**Review by danielguillan — APPROVED** [review]

**Pull request merged**
